# Assignment 6, Qilin Zhou, 02/15/2024

## Question 1

* The dataset is composed of 60,000 32×32-pixel color images (50,000 for training, 10,000 for testing) with 10 classes
Build a DNN with 20 hidden layers of 100 neurons each (that’s too many, but it’s the point of this exercise). Use He initialization and the ELU activation function. At the output layer use a softmax output layer with 10 neurons


In [1]:
import tensorflow as tf
from tensorflow import keras
import numpy as np

2024-02-15 23:53:19.099055: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


###  Load Data

In [4]:
(X_train_full, y_train_full), (X_test, y_test) = keras.datasets.cifar10.load_data()

### Build DNN

In [13]:
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
# The dataset is composed of 60,000 32×32-pixel color images 
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))

for i in range(20):
        model.add(keras.layers.Dense(100, activation="elu", kernel_initializer="he_normal"))

# Output layer
model.add(keras.layers.Dense(10, activation="softmax"))

## Question 2

### Using Nadam optimization and early stopping, train the network with Randomized Search for hyperparameter tuning

In [25]:
def build_model(learning_rate=0.001):
    model = keras.models.Sequential()
    model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
    
    for _ in range(20):
        model.add(keras.layers.Dense(100, activation="elu", kernel_initializer="he_normal"))

    model.add(keras.layers.Dense(10, activation="softmax"))

    optimizer = keras.optimizers.Nadam(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
    
    return model

from scikeras.wrappers import KerasClassifier

keras_classifier = KerasClassifier(build_fn=build_model)

In [27]:
from scipy.stats import reciprocal
from sklearn.model_selection import RandomizedSearchCV

param_distribs = { 
    "model__learning_rate": reciprocal(3e-5, 3e-2).rvs(1000).tolist(),
}

rnd_search_cv = RandomizedSearchCV(keras_classifier, param_distribs, n_iter=10, cv=3, verbose=2)

# Choose 3/4 of the training_full set as training
X_train = X_train_full[37500:]
y_train = y_train_full[37500:]
X_valid = X_train_full[:37500]
y_valid = y_train_full[:37500]

rnd_search_cv.fit(X_train, y_train, epochs=100,
                  validation_data=(X_valid, y_valid),
                  callbacks=[keras.callbacks.EarlyStopping(patience=10)])

Fitting 3 folds for each of 10 candidates, totalling 30 fits


/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 13s 20ms/step - loss: 4.4097 - accuracy: 0.1584 - val_loss: 2.2980 - val_accuracy: 0.1915
Epoch 2/100
261/261 [==============================] - 4s 17ms/step - loss: 2.1746 - accuracy: 0.2136 - val_loss: 2.1891 - val_accuracy: 0.2140
Epoch 3/100
261/261 [==============================] - 5s 18ms/step - loss: 2.0815 - accuracy: 0.2347 - val_loss: 2.0390 - val_accuracy: 0.2549
Epoch 4/100
261/261 [==============================] - 5s 18ms/step - loss: 2.0290 - accuracy: 0.2501 - val_loss: 1.9656 - val_accuracy: 0.2787
Epoch 5/100
261/261 [==============================] - 5s 18ms/step - loss: 1.9649 - accuracy: 0.2722 - val_loss: 2.0622 - val_accuracy: 0.2414
Epoch 6/100
261/261 [==============================] - 5s 19ms/step - loss: 1.9550 - accuracy: 0.2869 - val_loss: 2.0110 - val_accuracy: 0.2630
Epoch 7/100
261/261 [==============================] - 5s 20ms/step - loss: 1.9142 - accuracy: 0.2975 - val_loss: 1.9711 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 14s 21ms/step - loss: 6.0584 - accuracy: 0.1236 - val_loss: 2.3185 - val_accuracy: 0.2020
Epoch 2/100
261/261 [==============================] - 5s 19ms/step - loss: 2.2260 - accuracy: 0.1907 - val_loss: 2.1056 - val_accuracy: 0.2223
Epoch 3/100
261/261 [==============================] - 4s 17ms/step - loss: 2.1122 - accuracy: 0.2202 - val_loss: 2.0827 - val_accuracy: 0.2396
Epoch 4/100
261/261 [==============================] - 5s 18ms/step - loss: 2.0722 - accuracy: 0.2426 - val_loss: 2.0735 - val_accuracy: 0.2348
Epoch 5/100
261/261 [==============================] - 5s 18ms/step - loss: 2.0195 - accuracy: 0.2633 - val_loss: 1.9897 - val_accuracy: 0.2728
Epoch 6/100
261/261 [==============================] - 4s 17ms/step - loss: 1.9935 - accuracy: 0.2603 - val_loss: 2.0095 - val_accuracy: 0.2617
Epoch 7/100
261/261 [==============================] - 4s 17ms/step - loss: 1.9540 - accuracy: 0.2797 - val_loss: 1.9693 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 12s 19ms/step - loss: 5.5970 - accuracy: 0.1442 - val_loss: 2.1964 - val_accuracy: 0.1807
Epoch 2/100
261/261 [==============================] - 4s 16ms/step - loss: 2.1625 - accuracy: 0.2061 - val_loss: 2.1040 - val_accuracy: 0.2040
Epoch 3/100
261/261 [==============================] - 4s 17ms/step - loss: 2.0720 - accuracy: 0.2337 - val_loss: 2.0442 - val_accuracy: 0.2521
Epoch 4/100
261/261 [==============================] - 4s 16ms/step - loss: 2.0364 - accuracy: 0.2467 - val_loss: 2.0304 - val_accuracy: 0.2402
Epoch 5/100
261/261 [==============================] - 4s 16ms/step - loss: 1.9900 - accuracy: 0.2658 - val_loss: 1.9897 - val_accuracy: 0.2664
Epoch 6/100
261/261 [==============================] - 4s 16ms/step - loss: 1.9642 - accuracy: 0.2773 - val_loss: 1.9808 - val_accuracy: 0.2692
Epoch 7/100
261/261 [==============================] - 5s 18ms/step - loss: 1.9233 - accuracy: 0.2827 - val_loss: 1.9502 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 14s 21ms/step - loss: 112.3136 - accuracy: 0.1260 - val_loss: 2.2357 - val_accuracy: 0.1621
Epoch 2/100
261/261 [==============================] - 4s 17ms/step - loss: 2.1748 - accuracy: 0.1690 - val_loss: 2.1759 - val_accuracy: 0.1751
Epoch 3/100
261/261 [==============================] - 5s 18ms/step - loss: 2.1341 - accuracy: 0.1729 - val_loss: 2.0994 - val_accuracy: 0.1861
Epoch 4/100
261/261 [==============================] - 5s 18ms/step - loss: 2.1064 - accuracy: 0.1774 - val_loss: 2.0962 - val_accuracy: 0.1912
Epoch 5/100
261/261 [==============================] - 5s 18ms/step - loss: 2.0940 - accuracy: 0.1958 - val_loss: 2.0748 - val_accuracy: 0.1934
Epoch 6/100
261/261 [==============================] - 5s 17ms/step - loss: 2.0750 - accuracy: 0.1948 - val_loss: 2.1044 - val_accuracy: 0.1972
Epoch 7/100
261/261 [==============================] - 5s 19ms/step - loss: 2.0654 - accuracy: 0.2176 - val_loss: 2.0409 - val_accura

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 14s 25ms/step - loss: 173.1603 - accuracy: 0.1348 - val_loss: 2.3288 - val_accuracy: 0.1612
Epoch 2/100
261/261 [==============================] - 5s 20ms/step - loss: 2.2201 - accuracy: 0.1651 - val_loss: 2.1505 - val_accuracy: 0.1772
Epoch 3/100
261/261 [==============================] - 5s 18ms/step - loss: 2.2002 - accuracy: 0.1739 - val_loss: 2.3189 - val_accuracy: 0.1269
Epoch 4/100
261/261 [==============================] - 4s 16ms/step - loss: 2.1372 - accuracy: 0.1824 - val_loss: 2.1139 - val_accuracy: 0.1814
Epoch 5/100
261/261 [==============================] - 4s 17ms/step - loss: 2.1068 - accuracy: 0.1951 - val_loss: 2.1660 - val_accuracy: 0.1809
Epoch 6/100
261/261 [==============================] - 4s 16ms/step - loss: 2.0795 - accuracy: 0.2129 - val_loss: 2.0664 - val_accuracy: 0.1979
Epoch 7/100
261/261 [==============================] - 4s 16ms/step - loss: 2.0720 - accuracy: 0.2141 - val_loss: 2.1416 - val_accura

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 14s 21ms/step - loss: 78.8074 - accuracy: 0.1377 - val_loss: 2.2720 - val_accuracy: 0.1340
Epoch 2/100
261/261 [==============================] - 5s 20ms/step - loss: 2.1765 - accuracy: 0.1767 - val_loss: 2.1144 - val_accuracy: 0.1794
Epoch 3/100
261/261 [==============================] - 5s 19ms/step - loss: 2.1091 - accuracy: 0.1973 - val_loss: 2.2887 - val_accuracy: 0.1691
Epoch 4/100
261/261 [==============================] - 5s 18ms/step - loss: 2.0968 - accuracy: 0.1995 - val_loss: 2.0757 - val_accuracy: 0.1888
Epoch 5/100
261/261 [==============================] - 4s 17ms/step - loss: 2.0507 - accuracy: 0.2148 - val_loss: 2.0317 - val_accuracy: 0.2351
Epoch 6/100
261/261 [==============================] - 5s 18ms/step - loss: 2.0213 - accuracy: 0.2349 - val_loss: 2.0467 - val_accuracy: 0.2122
Epoch 7/100
261/261 [==============================] - 5s 19ms/step - loss: 2.0052 - accuracy: 0.2288 - val_loss: 1.9906 - val_accurac

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 14s 20ms/step - loss: 10.4885 - accuracy: 0.1195 - val_loss: 2.9302 - val_accuracy: 0.1553
Epoch 2/100
261/261 [==============================] - 6s 21ms/step - loss: 2.5579 - accuracy: 0.1764 - val_loss: 2.4751 - val_accuracy: 0.1848
Epoch 3/100
261/261 [==============================] - 6s 23ms/step - loss: 2.3011 - accuracy: 0.1991 - val_loss: 2.2538 - val_accuracy: 0.2062
Epoch 4/100
261/261 [==============================] - 5s 19ms/step - loss: 2.1723 - accuracy: 0.2254 - val_loss: 2.2478 - val_accuracy: 0.2169
Epoch 5/100
261/261 [==============================] - 5s 18ms/step - loss: 2.1002 - accuracy: 0.2520 - val_loss: 2.3895 - val_accuracy: 0.1993
Epoch 6/100
261/261 [==============================] - 5s 17ms/step - loss: 2.0536 - accuracy: 0.2550 - val_loss: 2.1282 - val_accuracy: 0.2372
Epoch 7/100
261/261 [==============================] - 4s 16ms/step - loss: 2.0143 - accuracy: 0.2736 - val_loss: 2.1850 - val_accurac

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 15s 22ms/step - loss: 8.9651 - accuracy: 0.1253 - val_loss: 3.1326 - val_accuracy: 0.1589
Epoch 2/100
261/261 [==============================] - 5s 20ms/step - loss: 2.5956 - accuracy: 0.1676 - val_loss: 2.3893 - val_accuracy: 0.1848
Epoch 3/100
261/261 [==============================] - 8s 30ms/step - loss: 2.2577 - accuracy: 0.2077 - val_loss: 2.2576 - val_accuracy: 0.2115
Epoch 4/100
261/261 [==============================] - 6s 21ms/step - loss: 2.1639 - accuracy: 0.2203 - val_loss: 2.1638 - val_accuracy: 0.2189
Epoch 5/100
261/261 [==============================] - 6s 21ms/step - loss: 2.0893 - accuracy: 0.2464 - val_loss: 2.1153 - val_accuracy: 0.2317
Epoch 6/100
261/261 [==============================] - 5s 21ms/step - loss: 2.0421 - accuracy: 0.2567 - val_loss: 2.1326 - val_accuracy: 0.2412
Epoch 7/100
261/261 [==============================] - 5s 21ms/step - loss: 2.0084 - accuracy: 0.2629 - val_loss: 2.1502 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 13s 20ms/step - loss: 7.3438 - accuracy: 0.1278 - val_loss: 2.7949 - val_accuracy: 0.1589
Epoch 2/100
261/261 [==============================] - 4s 17ms/step - loss: 2.4923 - accuracy: 0.1860 - val_loss: 2.3023 - val_accuracy: 0.1928
Epoch 3/100
261/261 [==============================] - 4s 17ms/step - loss: 2.2137 - accuracy: 0.2096 - val_loss: 2.2379 - val_accuracy: 0.2228
Epoch 4/100
261/261 [==============================] - 4s 17ms/step - loss: 2.1157 - accuracy: 0.2357 - val_loss: 2.0857 - val_accuracy: 0.2453
Epoch 5/100
261/261 [==============================] - 4s 16ms/step - loss: 2.0670 - accuracy: 0.2469 - val_loss: 2.0769 - val_accuracy: 0.2461
Epoch 6/100
261/261 [==============================] - 4s 16ms/step - loss: 2.0276 - accuracy: 0.2639 - val_loss: 2.0764 - val_accuracy: 0.2429
Epoch 7/100
261/261 [==============================] - 4s 17ms/step - loss: 1.9993 - accuracy: 0.2735 - val_loss: 2.0411 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 13s 21ms/step - loss: 29.3218 - accuracy: 0.1415 - val_loss: 2.2082 - val_accuracy: 0.1898
Epoch 2/100
261/261 [==============================] - 5s 18ms/step - loss: 2.1502 - accuracy: 0.1902 - val_loss: 2.1145 - val_accuracy: 0.1992
Epoch 3/100
261/261 [==============================] - 5s 18ms/step - loss: 2.0716 - accuracy: 0.2215 - val_loss: 2.0287 - val_accuracy: 0.2558
Epoch 4/100
261/261 [==============================] - 5s 18ms/step - loss: 2.0249 - accuracy: 0.2394 - val_loss: 1.9818 - val_accuracy: 0.2495
Epoch 5/100
261/261 [==============================] - 4s 17ms/step - loss: 1.9990 - accuracy: 0.2555 - val_loss: 2.0222 - val_accuracy: 0.2346
Epoch 6/100
261/261 [==============================] - 4s 17ms/step - loss: 1.9729 - accuracy: 0.2700 - val_loss: 1.9390 - val_accuracy: 0.2659
Epoch 7/100
261/261 [==============================] - 5s 18ms/step - loss: 1.9448 - accuracy: 0.2873 - val_loss: 1.9656 - val_accurac

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 17s 25ms/step - loss: 52.2435 - accuracy: 0.1276 - val_loss: 2.2877 - val_accuracy: 0.1524
Epoch 2/100
261/261 [==============================] - 6s 22ms/step - loss: 2.1969 - accuracy: 0.1830 - val_loss: 2.0887 - val_accuracy: 0.2116
Epoch 3/100
261/261 [==============================] - 5s 21ms/step - loss: 2.0897 - accuracy: 0.2174 - val_loss: 2.0344 - val_accuracy: 0.2182
Epoch 4/100
261/261 [==============================] - 5s 20ms/step - loss: 2.0723 - accuracy: 0.2155 - val_loss: 2.0310 - val_accuracy: 0.2365
Epoch 5/100
261/261 [==============================] - 5s 20ms/step - loss: 2.0219 - accuracy: 0.2370 - val_loss: 2.0429 - val_accuracy: 0.2190
Epoch 6/100
261/261 [==============================] - 5s 21ms/step - loss: 1.9887 - accuracy: 0.2511 - val_loss: 1.9930 - val_accuracy: 0.2522
Epoch 7/100
261/261 [==============================] - 5s 20ms/step - loss: 1.9794 - accuracy: 0.2580 - val_loss: 1.9763 - val_accurac

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 15s 25ms/step - loss: 30.1268 - accuracy: 0.1670 - val_loss: 2.1405 - val_accuracy: 0.1962
Epoch 2/100
261/261 [==============================] - 5s 21ms/step - loss: 2.0756 - accuracy: 0.2161 - val_loss: 2.0832 - val_accuracy: 0.2093
Epoch 3/100
261/261 [==============================] - 5s 20ms/step - loss: 2.0289 - accuracy: 0.2433 - val_loss: 2.0233 - val_accuracy: 0.2335
Epoch 4/100
261/261 [==============================] - 6s 21ms/step - loss: 1.9850 - accuracy: 0.2634 - val_loss: 1.9904 - val_accuracy: 0.2506
Epoch 5/100
261/261 [==============================] - 5s 21ms/step - loss: 1.9661 - accuracy: 0.2664 - val_loss: 1.9415 - val_accuracy: 0.2729
Epoch 6/100
261/261 [==============================] - 6s 23ms/step - loss: 1.9329 - accuracy: 0.2822 - val_loss: 1.9448 - val_accuracy: 0.2601
Epoch 7/100
261/261 [==============================] - 6s 24ms/step - loss: 1.9243 - accuracy: 0.2819 - val_loss: 1.9020 - val_accurac

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 13s 19ms/step - loss: 7.6349 - accuracy: 0.1350 - val_loss: 2.3421 - val_accuracy: 0.1786
Epoch 2/100
261/261 [==============================] - 4s 16ms/step - loss: 2.2781 - accuracy: 0.1760 - val_loss: 2.3456 - val_accuracy: 0.1836
Epoch 3/100
261/261 [==============================] - 4s 17ms/step - loss: 2.1135 - accuracy: 0.2222 - val_loss: 2.1324 - val_accuracy: 0.2215
Epoch 4/100
261/261 [==============================] - 4s 17ms/step - loss: 2.0571 - accuracy: 0.2413 - val_loss: 2.0472 - val_accuracy: 0.2395
Epoch 5/100
261/261 [==============================] - 4s 16ms/step - loss: 2.0211 - accuracy: 0.2589 - val_loss: 2.1905 - val_accuracy: 0.2048
Epoch 6/100
261/261 [==============================] - 4s 17ms/step - loss: 1.9670 - accuracy: 0.2755 - val_loss: 2.0203 - val_accuracy: 0.2523
Epoch 7/100
261/261 [==============================] - 4s 17ms/step - loss: 1.9232 - accuracy: 0.2929 - val_loss: 1.9807 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 13s 19ms/step - loss: 5.0766 - accuracy: 0.1586 - val_loss: 2.2710 - val_accuracy: 0.2097
Epoch 2/100
261/261 [==============================] - 4s 16ms/step - loss: 2.1642 - accuracy: 0.2128 - val_loss: 2.1138 - val_accuracy: 0.2342
Epoch 3/100
261/261 [==============================] - 4s 17ms/step - loss: 2.0625 - accuracy: 0.2466 - val_loss: 2.0254 - val_accuracy: 0.2569
Epoch 4/100
261/261 [==============================] - 4s 17ms/step - loss: 2.0338 - accuracy: 0.2541 - val_loss: 2.0350 - val_accuracy: 0.2584
Epoch 5/100
261/261 [==============================] - 4s 16ms/step - loss: 1.9904 - accuracy: 0.2707 - val_loss: 2.0121 - val_accuracy: 0.2620
Epoch 6/100
261/261 [==============================] - 4s 17ms/step - loss: 1.9224 - accuracy: 0.2946 - val_loss: 1.9749 - val_accuracy: 0.2731
Epoch 7/100
261/261 [==============================] - 5s 17ms/step - loss: 1.9047 - accuracy: 0.3024 - val_loss: 1.9560 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 1642s 40ms/step - loss: 6.0483 - accuracy: 0.1547 - val_loss: 2.2557 - val_accuracy: 0.1806
Epoch 2/100
261/261 [==============================] - 13s 50ms/step - loss: 2.1542 - accuracy: 0.2120 - val_loss: 2.1175 - val_accuracy: 0.2074
Epoch 3/100
261/261 [==============================] - 13s 49ms/step - loss: 2.0688 - accuracy: 0.2465 - val_loss: 2.1926 - val_accuracy: 0.2229
Epoch 4/100
261/261 [==============================] - 1037s 4s/step - loss: 2.0241 - accuracy: 0.2630 - val_loss: 1.9863 - val_accuracy: 0.2650
Epoch 5/100
261/261 [==============================] - 5s 17ms/step - loss: 1.9738 - accuracy: 0.2784 - val_loss: 1.9793 - val_accuracy: 0.2682
Epoch 6/100
261/261 [==============================] - 6s 22ms/step - loss: 1.9541 - accuracy: 0.2852 - val_loss: 2.0004 - val_accuracy: 0.2710
Epoch 7/100
261/261 [==============================] - 13s 51ms/step - loss: 1.9375 - accuracy: 0.2951 - val_loss: 1.9614 - val_ac

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 23s 35ms/step - loss: 611.9146 - accuracy: 0.1333 - val_loss: 2.2708 - val_accuracy: 0.1846
Epoch 2/100
261/261 [==============================] - 8s 29ms/step - loss: 2.1583 - accuracy: 0.1896 - val_loss: 2.1605 - val_accuracy: 0.1796
Epoch 3/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0903 - accuracy: 0.2114 - val_loss: 2.0345 - val_accuracy: 0.2382
Epoch 4/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0762 - accuracy: 0.2131 - val_loss: 2.0765 - val_accuracy: 0.2054
Epoch 5/100
261/261 [==============================] - 7s 29ms/step - loss: 2.0562 - accuracy: 0.2252 - val_loss: 2.1048 - val_accuracy: 0.1924
Epoch 6/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0299 - accuracy: 0.2310 - val_loss: 2.0235 - val_accuracy: 0.2247
Epoch 7/100
261/261 [==============================] - 8s 29ms/step - loss: 2.0058 - accuracy: 0.2419 - val_loss: 2.0312 - val_accura

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 23s 35ms/step - loss: 129.1908 - accuracy: 0.0972 - val_loss: 2.3720 - val_accuracy: 0.1002
Epoch 2/100
261/261 [==============================] - 8s 30ms/step - loss: 2.3343 - accuracy: 0.1002 - val_loss: 2.3466 - val_accuracy: 0.1017
Epoch 3/100
261/261 [==============================] - 8s 30ms/step - loss: 2.3331 - accuracy: 0.1036 - val_loss: 2.3263 - val_accuracy: 0.0998
Epoch 4/100
261/261 [==============================] - 8s 30ms/step - loss: 2.3290 - accuracy: 0.0972 - val_loss: 2.3421 - val_accuracy: 0.1004
Epoch 5/100
261/261 [==============================] - 8s 30ms/step - loss: 2.3287 - accuracy: 0.1016 - val_loss: 2.3310 - val_accuracy: 0.0994
Epoch 6/100
261/261 [==============================] - 8s 31ms/step - loss: 2.3295 - accuracy: 0.0994 - val_loss: 2.3533 - val_accuracy: 0.0992
Epoch 7/100
261/261 [==============================] - 8s 31ms/step - loss: 2.3316 - accuracy: 0.1019 - val_loss: 2.3619 - val_accura

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 24s 36ms/step - loss: 307.7269 - accuracy: 0.1316 - val_loss: 2.2631 - val_accuracy: 0.1481
Epoch 2/100
261/261 [==============================] - 8s 31ms/step - loss: 2.1982 - accuracy: 0.1737 - val_loss: 2.1979 - val_accuracy: 0.1542
Epoch 3/100
261/261 [==============================] - 8s 30ms/step - loss: 2.1112 - accuracy: 0.1982 - val_loss: 2.1362 - val_accuracy: 0.2106
Epoch 4/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0817 - accuracy: 0.2149 - val_loss: 2.0879 - val_accuracy: 0.2018
Epoch 5/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0625 - accuracy: 0.2175 - val_loss: 2.0967 - val_accuracy: 0.2079
Epoch 6/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0576 - accuracy: 0.2201 - val_loss: 2.0929 - val_accuracy: 0.2185
Epoch 7/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0397 - accuracy: 0.2214 - val_loss: 2.0201 - val_accura

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 22s 34ms/step - loss: 5.7191 - accuracy: 0.1550 - val_loss: 2.4931 - val_accuracy: 0.1871
Epoch 2/100
261/261 [==============================] - 7s 28ms/step - loss: 2.2374 - accuracy: 0.2053 - val_loss: 2.2293 - val_accuracy: 0.1959
Epoch 3/100
261/261 [==============================] - 7s 28ms/step - loss: 2.1239 - accuracy: 0.2267 - val_loss: 2.0843 - val_accuracy: 0.2369
Epoch 4/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0570 - accuracy: 0.2452 - val_loss: 2.1163 - val_accuracy: 0.2384
Epoch 5/100
261/261 [==============================] - 8s 29ms/step - loss: 2.0316 - accuracy: 0.2561 - val_loss: 2.0940 - val_accuracy: 0.2433
Epoch 6/100
261/261 [==============================] - 8s 30ms/step - loss: 1.9649 - accuracy: 0.2752 - val_loss: 1.9895 - val_accuracy: 0.2667
Epoch 7/100
261/261 [==============================] - 8s 30ms/step - loss: 1.9365 - accuracy: 0.2899 - val_loss: 1.9496 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 22s 32ms/step - loss: 4.6742 - accuracy: 0.1528 - val_loss: 2.3583 - val_accuracy: 0.1796
Epoch 2/100
261/261 [==============================] - 7s 27ms/step - loss: 2.1977 - accuracy: 0.2107 - val_loss: 2.1707 - val_accuracy: 0.2049
Epoch 3/100
261/261 [==============================] - 7s 27ms/step - loss: 2.0737 - accuracy: 0.2443 - val_loss: 2.0688 - val_accuracy: 0.2428
Epoch 4/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0243 - accuracy: 0.2575 - val_loss: 2.0109 - val_accuracy: 0.2530
Epoch 5/100
261/261 [==============================] - 7s 26ms/step - loss: 1.9906 - accuracy: 0.2691 - val_loss: 2.0037 - val_accuracy: 0.2668
Epoch 6/100
261/261 [==============================] - 7s 27ms/step - loss: 1.9336 - accuracy: 0.2867 - val_loss: 1.9298 - val_accuracy: 0.2869
Epoch 7/100
261/261 [==============================] - 7s 27ms/step - loss: 1.9122 - accuracy: 0.3012 - val_loss: 1.9668 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 21s 31ms/step - loss: 9.5990 - accuracy: 0.1404 - val_loss: 2.2446 - val_accuracy: 0.1705
Epoch 2/100
261/261 [==============================] - 7s 26ms/step - loss: 2.1791 - accuracy: 0.2021 - val_loss: 2.1559 - val_accuracy: 0.1993
Epoch 3/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0954 - accuracy: 0.2285 - val_loss: 2.1782 - val_accuracy: 0.2181
Epoch 4/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0434 - accuracy: 0.2471 - val_loss: 2.0195 - val_accuracy: 0.2479
Epoch 5/100
261/261 [==============================] - 7s 25ms/step - loss: 2.0048 - accuracy: 0.2648 - val_loss: 1.9895 - val_accuracy: 0.2669
Epoch 6/100
261/261 [==============================] - 7s 25ms/step - loss: 1.9613 - accuracy: 0.2789 - val_loss: 2.0468 - val_accuracy: 0.2467
Epoch 7/100
261/261 [==============================] - 7s 25ms/step - loss: 1.9403 - accuracy: 0.2888 - val_loss: 1.9679 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 21s 31ms/step - loss: 7.7145 - accuracy: 0.1692 - val_loss: 2.1364 - val_accuracy: 0.2109
Epoch 2/100
261/261 [==============================] - 7s 26ms/step - loss: 2.1090 - accuracy: 0.2166 - val_loss: 2.1240 - val_accuracy: 0.2097
Epoch 3/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0380 - accuracy: 0.2431 - val_loss: 1.9915 - val_accuracy: 0.2622
Epoch 4/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0013 - accuracy: 0.2536 - val_loss: 1.9865 - val_accuracy: 0.2556
Epoch 5/100
261/261 [==============================] - 7s 25ms/step - loss: 1.9624 - accuracy: 0.2743 - val_loss: 2.1136 - val_accuracy: 0.2310
Epoch 6/100
261/261 [==============================] - 7s 26ms/step - loss: 1.9237 - accuracy: 0.2867 - val_loss: 1.9341 - val_accuracy: 0.2835
Epoch 7/100
261/261 [==============================] - 7s 26ms/step - loss: 1.8808 - accuracy: 0.3064 - val_loss: 1.9051 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 22s 33ms/step - loss: 11.5573 - accuracy: 0.1615 - val_loss: 2.1671 - val_accuracy: 0.2138
Epoch 2/100
261/261 [==============================] - 7s 28ms/step - loss: 2.1337 - accuracy: 0.2150 - val_loss: 2.1133 - val_accuracy: 0.2061
Epoch 3/100
261/261 [==============================] - 7s 27ms/step - loss: 2.0656 - accuracy: 0.2297 - val_loss: 2.0327 - val_accuracy: 0.2425
Epoch 4/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0165 - accuracy: 0.2480 - val_loss: 1.9794 - val_accuracy: 0.2735
Epoch 5/100
261/261 [==============================] - 7s 28ms/step - loss: 1.9872 - accuracy: 0.2595 - val_loss: 1.9873 - val_accuracy: 0.2561
Epoch 6/100
261/261 [==============================] - 7s 27ms/step - loss: 1.9582 - accuracy: 0.2753 - val_loss: 1.9475 - val_accuracy: 0.2635
Epoch 7/100
261/261 [==============================] - 7s 27ms/step - loss: 1.9519 - accuracy: 0.2783 - val_loss: 1.9670 - val_accurac

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 22s 34ms/step - loss: 11.7738 - accuracy: 0.1363 - val_loss: 2.1716 - val_accuracy: 0.1908
Epoch 2/100
261/261 [==============================] - 8s 32ms/step - loss: 2.1225 - accuracy: 0.2023 - val_loss: 2.0814 - val_accuracy: 0.2188
Epoch 3/100
261/261 [==============================] - 8s 30ms/step - loss: 2.0680 - accuracy: 0.2274 - val_loss: 2.0482 - val_accuracy: 0.2501
Epoch 4/100
261/261 [==============================] - 8s 30ms/step - loss: 2.0077 - accuracy: 0.2477 - val_loss: 2.0816 - val_accuracy: 0.2403
Epoch 5/100
261/261 [==============================] - 8s 30ms/step - loss: 1.9803 - accuracy: 0.2652 - val_loss: 2.0006 - val_accuracy: 0.2498
Epoch 6/100
261/261 [==============================] - 8s 29ms/step - loss: 1.9544 - accuracy: 0.2747 - val_loss: 1.9712 - val_accuracy: 0.2682
Epoch 7/100
261/261 [==============================] - 8s 29ms/step - loss: 1.9193 - accuracy: 0.2953 - val_loss: 1.9165 - val_accurac

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 22s 32ms/step - loss: 7.3512 - accuracy: 0.1583 - val_loss: 2.3045 - val_accuracy: 0.1787
Epoch 2/100
261/261 [==============================] - 7s 27ms/step - loss: 2.1423 - accuracy: 0.2040 - val_loss: 2.1192 - val_accuracy: 0.2173
Epoch 3/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0669 - accuracy: 0.2270 - val_loss: 2.0583 - val_accuracy: 0.2359
Epoch 4/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0182 - accuracy: 0.2524 - val_loss: 2.0027 - val_accuracy: 0.2488
Epoch 5/100
261/261 [==============================] - 7s 26ms/step - loss: 1.9844 - accuracy: 0.2625 - val_loss: 2.1294 - val_accuracy: 0.2295
Epoch 6/100
261/261 [==============================] - 7s 26ms/step - loss: 1.9488 - accuracy: 0.2733 - val_loss: 1.9981 - val_accuracy: 0.2557
Epoch 7/100
261/261 [==============================] - 7s 26ms/step - loss: 1.9445 - accuracy: 0.2865 - val_loss: 1.9435 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 23s 32ms/step - loss: 5.2128 - accuracy: 0.1372 - val_loss: 2.3288 - val_accuracy: 0.1882
Epoch 2/100
261/261 [==============================] - 7s 26ms/step - loss: 2.1868 - accuracy: 0.1967 - val_loss: 2.1071 - val_accuracy: 0.2014
Epoch 3/100
261/261 [==============================] - 7s 27ms/step - loss: 2.0848 - accuracy: 0.2268 - val_loss: 2.0796 - val_accuracy: 0.2206
Epoch 4/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0333 - accuracy: 0.2477 - val_loss: 2.0261 - val_accuracy: 0.2418
Epoch 5/100
261/261 [==============================] - 7s 26ms/step - loss: 1.9831 - accuracy: 0.2692 - val_loss: 2.0110 - val_accuracy: 0.2573
Epoch 6/100
261/261 [==============================] - 7s 26ms/step - loss: 1.9516 - accuracy: 0.2687 - val_loss: 1.9415 - val_accuracy: 0.2821
Epoch 7/100
261/261 [==============================] - 7s 27ms/step - loss: 1.9202 - accuracy: 0.2872 - val_loss: 1.9766 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 22s 34ms/step - loss: 6.5833 - accuracy: 0.1441 - val_loss: 2.2714 - val_accuracy: 0.1815
Epoch 2/100
261/261 [==============================] - 7s 28ms/step - loss: 2.2010 - accuracy: 0.1995 - val_loss: 2.1420 - val_accuracy: 0.2132
Epoch 3/100
261/261 [==============================] - 7s 27ms/step - loss: 2.0916 - accuracy: 0.2390 - val_loss: 2.2777 - val_accuracy: 0.2170
Epoch 4/100
261/261 [==============================] - 7s 27ms/step - loss: 2.0334 - accuracy: 0.2546 - val_loss: 2.0220 - val_accuracy: 0.2371
Epoch 5/100
261/261 [==============================] - 7s 28ms/step - loss: 1.9906 - accuracy: 0.2690 - val_loss: 1.9852 - val_accuracy: 0.2684
Epoch 6/100
261/261 [==============================] - 7s 27ms/step - loss: 1.9332 - accuracy: 0.2881 - val_loss: 1.9339 - val_accuracy: 0.2878
Epoch 7/100
261/261 [==============================] - 7s 27ms/step - loss: 1.9059 - accuracy: 0.2963 - val_loss: 1.9616 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 20s 30ms/step - loss: 7.6157 - accuracy: 0.1224 - val_loss: 2.5693 - val_accuracy: 0.1509
Epoch 2/100
261/261 [==============================] - 7s 25ms/step - loss: 2.3718 - accuracy: 0.1699 - val_loss: 2.2592 - val_accuracy: 0.1942
Epoch 3/100
261/261 [==============================] - 7s 25ms/step - loss: 2.1724 - accuracy: 0.2160 - val_loss: 2.1532 - val_accuracy: 0.2274
Epoch 4/100
261/261 [==============================] - 7s 25ms/step - loss: 2.0844 - accuracy: 0.2458 - val_loss: 2.0745 - val_accuracy: 0.2483
Epoch 5/100
261/261 [==============================] - 6s 25ms/step - loss: 2.0403 - accuracy: 0.2634 - val_loss: 2.0805 - val_accuracy: 0.2320
Epoch 6/100
261/261 [==============================] - 7s 25ms/step - loss: 1.9867 - accuracy: 0.2853 - val_loss: 2.0399 - val_accuracy: 0.2530
Epoch 7/100
261/261 [==============================] - 6s 25ms/step - loss: 1.9703 - accuracy: 0.2837 - val_loss: 2.0755 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 22s 34ms/step - loss: 9.1949 - accuracy: 0.1334 - val_loss: 2.6850 - val_accuracy: 0.1613
Epoch 2/100
261/261 [==============================] - 7s 28ms/step - loss: 2.3548 - accuracy: 0.1859 - val_loss: 2.2584 - val_accuracy: 0.1941
Epoch 3/100
261/261 [==============================] - 7s 28ms/step - loss: 2.1708 - accuracy: 0.2227 - val_loss: 2.1771 - val_accuracy: 0.2222
Epoch 4/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0986 - accuracy: 0.2432 - val_loss: 2.1275 - val_accuracy: 0.2176
Epoch 5/100
261/261 [==============================] - 7s 28ms/step - loss: 2.0505 - accuracy: 0.2507 - val_loss: 2.0788 - val_accuracy: 0.2416
Epoch 6/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0124 - accuracy: 0.2604 - val_loss: 2.0154 - val_accuracy: 0.2633
Epoch 7/100
261/261 [==============================] - 7s 25ms/step - loss: 1.9766 - accuracy: 0.2764 - val_loss: 2.0325 - val_accuracy

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
261/261 [==============================] - 21s 31ms/step - loss: 11.3332 - accuracy: 0.1224 - val_loss: 2.7420 - val_accuracy: 0.1481
Epoch 2/100
261/261 [==============================] - 7s 26ms/step - loss: 2.4605 - accuracy: 0.1673 - val_loss: 2.3159 - val_accuracy: 0.1782
Epoch 3/100
261/261 [==============================] - 7s 26ms/step - loss: 2.2259 - accuracy: 0.1976 - val_loss: 2.2101 - val_accuracy: 0.2030
Epoch 4/100
261/261 [==============================] - 7s 26ms/step - loss: 2.1259 - accuracy: 0.2208 - val_loss: 2.0688 - val_accuracy: 0.2301
Epoch 5/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0691 - accuracy: 0.2412 - val_loss: 2.1047 - val_accuracy: 0.2216
Epoch 6/100
261/261 [==============================] - 7s 26ms/step - loss: 2.0327 - accuracy: 0.2510 - val_loss: 2.0326 - val_accuracy: 0.2523
Epoch 7/100
261/261 [==============================] - 7s 26ms/step - loss: 1.9846 - accuracy: 0.2714 - val_loss: 2.0323 - val_accurac

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


Epoch 1/100
391/391 [==============================] - 24s 27ms/step - loss: 5.4214 - accuracy: 0.1706 - val_loss: 2.1093 - val_accuracy: 0.2274
Epoch 2/100
391/391 [==============================] - 9s 24ms/step - loss: 2.0906 - accuracy: 0.2312 - val_loss: 2.0447 - val_accuracy: 0.2350
Epoch 3/100
391/391 [==============================] - 9s 23ms/step - loss: 2.0267 - accuracy: 0.2501 - val_loss: 2.0111 - val_accuracy: 0.2455
Epoch 4/100
391/391 [==============================] - 9s 23ms/step - loss: 1.9834 - accuracy: 0.2686 - val_loss: 1.9709 - val_accuracy: 0.2808
Epoch 5/100
391/391 [==============================] - 10s 25ms/step - loss: 1.9232 - accuracy: 0.2940 - val_loss: 1.8778 - val_accuracy: 0.3170
Epoch 6/100
391/391 [==============================] - 10s 27ms/step - loss: 1.8959 - accuracy: 0.3007 - val_loss: 1.8924 - val_accuracy: 0.3001
Epoch 7/100
391/391 [==============================] - 10s 27ms/step - loss: 1.8629 - accuracy: 0.3181 - val_loss: 1.8650 - val_accur

RandomizedSearchCV(cv=3,
                   estimator=KerasClassifier(build_fn=<function build_model at 0x15b7a5080>),
                   param_distributions={'model__learning_rate': [0.01192544089977353,
                                                                 0.001095892084442962,
                                                                 0.00305434887884912,
                                                                 0.008359247899717575,
                                                                 0.00018161291448501717,
                                                                 0.0002332428167492523,
                                                                 3.0130324195205462e-05,
                                                                 0.023248966270150648,
                                                                 0.016972895545961188,
                                                                 0.0058996472607239...
                                                                 0.0021167528362478283,
                                                                 0.022888087546062717,
                                                                 0.015336467019355123,
                                                                 0.0036930543533677876,
                                                                 5.848479974701813e-05,
                                                                 0.00019934868157775343,
                                                                 0.0002737104857384861,
                                                                 0.005842132841357289,
                                                                 0.007653653423987863,
                                                                 0.0004289825472562583,
                                                                 0.0005379888124804675,
                                                                 4.079889234703265e-05,
                                                                 0.0002210708138087823,
                                                                 0.0006776052875363995,
                                                                 6.730857464902144e-05,
                                                                 0.0006869476417826836, ...]},
                   verbose=2)

In [29]:
rnd_search_cv.best_params_

{'model__learning_rate': 0.00042199648026968004}

* The best tuned learning rate is thus set to be 0.000422

In [34]:
model = build_model(learning_rate=0.000422)

In [37]:
model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 2ms/step - loss: 279.0726 - accuracy: 0.1097


[279.0726013183594, 0.10970000177621841]

* The current model accuracy is around 10.97%.

## Question 3

### Add Batch Normalization and compare the learning curves

In [30]:
import os
root_logdir = os.path.join(os.curdir, "my_cifar10_logs")

def get_run_logdir():
    import time
    run_id = time.strftime("run_%Y_%m_%d-%H_%M_%S")
    return os.path.join(root_logdir, run_id)

run_logdir = get_run_logdir()

In [38]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
model.add(keras.layers.BatchNormalization())
for _ in range(20):
    model.add(keras.layers.Dense(100, kernel_initializer="he_normal"))
    model.add(keras.layers.BatchNormalization())
    model.add(keras.layers.Activation("elu"))
model.add(keras.layers.Dense(10, activation="softmax"))

optimizer = keras.optimizers.Nadam(learning_rate=0.000422) 
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)
checkpoint_cb = keras.callbacks.ModelCheckpoint("my_cifar10_bn_model.h5",
                                                save_best_only=True)

early_stopping_cb = keras.callbacks.EarlyStopping(patience=10,
                                                  restore_best_weights=True)

model.fit(X_train, y_train, epochs=100,
                    validation_data=(X_valid, y_valid),
                    callbacks=[checkpoint_cb, early_stopping_cb, tensorboard_cb])

model = keras.models.load_model("my_cifar10_bn_model.h5")
model.evaluate(X_test, y_test)

Epoch 1/100
388/391 [============================>.] - ETA: 0s - loss: 2.0347 - accuracy: 0.2724

/Users/qilinzhou/anaconda3/lib/python3.11/site-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


391/391 [==============================] - 30s 27ms/step - loss: 2.0346 - accuracy: 0.2726 - val_loss: 1.8385 - val_accuracy: 0.3398
Epoch 2/100
391/391 [==============================] - 9s 22ms/step - loss: 1.8096 - accuracy: 0.3479 - val_loss: 1.7461 - val_accuracy: 0.3697
Epoch 3/100
391/391 [==============================] - 9s 24ms/step - loss: 1.7163 - accuracy: 0.3832 - val_loss: 1.7019 - val_accuracy: 0.3878
Epoch 4/100
391/391 [==============================] - 10s 25ms/step - loss: 1.6570 - accuracy: 0.4109 - val_loss: 1.6471 - val_accuracy: 0.4141
Epoch 5/100
391/391 [==============================] - 10s 26ms/step - loss: 1.6091 - accuracy: 0.4320 - val_loss: 1.6303 - val_accuracy: 0.4215
Epoch 6/100
391/391 [==============================] - 9s 22ms/step - loss: 1.5664 - accuracy: 0.4429 - val_loss: 1.6357 - val_accuracy: 0.4199
Epoch 7/100
391/391 [==============================] - 9s 22ms/step - loss: 1.5365 - accuracy: 0.4590 - val_loss: 1.6212 - val_accuracy: 0.4234
E

[1.5583374500274658, 0.4542999863624573]

In [47]:
%reload_ext tensorboard
%tensorboard --logdir=./my_cifar10_logs --port=6006 --host=0.0.0.0

Reusing TensorBoard on port 6006 (pid 2213), started 0:01:33 ago. (Use '!kill 2213' to kill it.)

* Batch Normalization makes the model converge faster, taking only 20 epochs to reach the lowest validation loss, with test accuracy around 45.43%. This is a significant improvement from the prior model with only 10.97% accuracy. The training speed is not faster, with around 9s for each epoch, while basline model has around 7s for each epoch, on average.

## Question 4

###  Replacing Batch Normalization with SELU

* standardize the input features

In [48]:
means = X_train.mean(axis=0)
stds = X_train.std(axis=0)
X_train_scaled = (X_train - means) / stds
X_valid_scaled = (X_valid - means) / stds
X_test_scaled = (X_test - means) / stds

* use LeCun normal initialization, make sure the DNN contains only a sequence of dense layers

In [52]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
for _ in range(20):
    model.add(keras.layers.Dense(100,
                                 kernel_initializer="lecun_normal",
                                 activation="selu"))
model.add(keras.layers.Dense(10, activation="softmax"))

optimizer = keras.optimizers.Nadam(learning_rate=0.000422) 
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

In [53]:
tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)
checkpoint_cb = keras.callbacks.ModelCheckpoint("my_cifar10_selu_model.h5",
                                                save_best_only=True)

early_stopping_cb = keras.callbacks.EarlyStopping(patience=10,
                                                  restore_best_weights=True)

model.fit(X_train_scaled, y_train, epochs=100,
          validation_data=(X_valid_scaled, y_valid),
          callbacks=[checkpoint_cb, early_stopping_cb, tensorboard_cb])

model = keras.models.load_model("my_cifar10_selu_model.h5")
model.evaluate(X_test_scaled, y_test)

Epoch 1/100
391/391 [==============================] - 19s 24ms/step - loss: 2.0484 - accuracy: 0.2646 - val_loss: 1.8849 - val_accuracy: 0.3237
Epoch 2/100
391/391 [==============================] - 6s 16ms/step - loss: 1.8558 - accuracy: 0.3382 - val_loss: 1.8375 - val_accuracy: 0.3391
Epoch 3/100
391/391 [==============================] - 6s 16ms/step - loss: 1.7538 - accuracy: 0.3695 - val_loss: 1.7595 - val_accuracy: 0.3669
Epoch 4/100
391/391 [==============================] - 6s 16ms/step - loss: 1.6596 - accuracy: 0.4082 - val_loss: 1.7577 - val_accuracy: 0.3809
Epoch 5/100
391/391 [==============================] - 6s 16ms/step - loss: 1.6027 - accuracy: 0.4314 - val_loss: 1.7202 - val_accuracy: 0.3973
Epoch 6/100
391/391 [==============================] - 6s 16ms/step - loss: 1.5226 - accuracy: 0.4626 - val_loss: 1.7082 - val_accuracy: 0.4054
Epoch 7/100
391/391 [==============================] - 6s 16ms/step - loss: 1.4634 - accuracy: 0.4854 - val_loss: 1.6946 - val_accuracy

[1.6753941774368286, 0.4250999987125397]

In [54]:
%tensorboard --logdir=./my_cifar10_logs --port=6006 --host=0.0.0.0

Reusing TensorBoard on port 6006 (pid 2213), started 0:18:19 ago. (Use '!kill 2213' to kill it.)

* SELU method converges faster, taking only 17 epochs to reach the lowest validation loss. However, it does not improve the test accuracy from Batch Normalization, with 42.51% test accuracy.

## Question 5

### Try regularizing the model with alpha dropout. 

In [57]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
for _ in range(20):
    model.add(keras.layers.AlphaDropout(rate=0.2))
    model.add(keras.layers.Dense(100,
                                 kernel_initializer="lecun_normal",
                                 activation="selu"))

model.add(keras.layers.Dense(10, activation="softmax"))

optimizer = keras.optimizers.Nadam(learning_rate=0.000422) 
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

In [59]:
tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)
checkpoint_cb = keras.callbacks.ModelCheckpoint("my_cifar10_alpha_model.h5",
                                                save_best_only=True)

early_stopping_cb = keras.callbacks.EarlyStopping(patience=20,
                                                  restore_best_weights=True)

model.fit(X_train_scaled, y_train, epochs=100,
          validation_data=(X_valid_scaled, y_valid),
          callbacks=[checkpoint_cb, early_stopping_cb, tensorboard_cb])

model = keras.models.load_model("my_cifar10_alpha_model.h5")
model.evaluate(X_test_scaled, y_test)

Epoch 1/100
391/391 [==============================] - 17s 28ms/step - loss: 2.2136 - accuracy: 0.1725 - val_loss: 3.4971 - val_accuracy: 0.1512
Epoch 2/100
391/391 [==============================] - 11s 28ms/step - loss: 2.1885 - accuracy: 0.1643 - val_loss: 5.6947 - val_accuracy: 0.1810
Epoch 3/100
391/391 [==============================] - 10s 26ms/step - loss: 2.1484 - accuracy: 0.1589 - val_loss: 20.0929 - val_accuracy: 0.1577
Epoch 4/100
391/391 [==============================] - 10s 25ms/step - loss: 2.1233 - accuracy: 0.1669 - val_loss: 16.0944 - val_accuracy: 0.1642
Epoch 5/100
391/391 [==============================] - 11s 28ms/step - loss: 2.0941 - accuracy: 0.1747 - val_loss: 25.3057 - val_accuracy: 0.1675
Epoch 6/100
391/391 [==============================] - 11s 28ms/step - loss: 2.0868 - accuracy: 0.1749 - val_loss: 32.7706 - val_accuracy: 0.1279
Epoch 7/100
391/391 [==============================] - 11s 28ms/step - loss: 2.0696 - accuracy: 0.1822 - val_loss: 23.0864 - v

[3.5354628562927246, 0.15150000154972076]

In [60]:
%tensorboard --logdir=./my_cifar10_logs --port=6006 --host=0.0.0.0

Reusing TensorBoard on port 6006 (pid 2213), started 0:35:09 ago. (Use '!kill 2213' to kill it.)

The model with alpha dropout increases the baseline model test accuracy from 10.97% to 15.15%.

### using MC Dropout

In [61]:
class MCAlphaDropout(keras.layers.AlphaDropout):
    def call(self, inputs):
        return super().call(inputs, training=True)

In [63]:
mc_model = keras.models.Sequential([
    MCAlphaDropout(layer.rate) if isinstance(layer, keras.layers.AlphaDropout) else layer
    for layer in model.layers
])

In [64]:
y_probas = np.stack([model(X_test_scaled, training=True)
                     for sample in range(100)])
y_proba = y_probas.mean(axis=0)
y_pred = np.argmax(y_proba, axis=1)

In [65]:
accuracy = np.mean(y_pred == y_test[:, 0])
accuracy

0.1672

* We achieved higher acccuracy from MC Dropout, which is now 16.72%,  compared with Alpha Dropout (15.15%).

## Question 6

### 1cycle scheduling

In [66]:
import math
K = keras.backend

class OneCycleScheduler(keras.callbacks.Callback):
    def __init__(self, iterations, max_rate, start_rate=None,
                 last_iterations=None, last_rate=None):
        self.iterations = iterations
        self.max_rate = max_rate
        self.start_rate = start_rate or max_rate / 10
        self.last_iterations = last_iterations or iterations // 10 + 1
        self.half_iteration = (iterations - self.last_iterations) // 2
        self.last_rate = last_rate or self.start_rate / 1000
        self.iteration = 0
    def _interpolate(self, iter1, iter2, rate1, rate2):
        return ((rate2 - rate1) * (self.iteration - iter1)
                / (iter2 - iter1) + rate1)
    def on_batch_begin(self, batch, logs):
        if self.iteration < self.half_iteration:
            rate = self._interpolate(0, self.half_iteration, self.start_rate, self.max_rate)
        elif self.iteration < 2 * self.half_iteration:
            rate = self._interpolate(self.half_iteration, 2 * self.half_iteration,
                                     self.max_rate, self.start_rate)
        else:
            rate = self._interpolate(2 * self.half_iteration, self.iterations,
                                     self.start_rate, self.last_rate)
        self.iteration += 1
        K.set_value(self.model.optimizer.lr, rate)

In [70]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
for _ in range(20):
    model.add(keras.layers.Dense(100,
                                 kernel_initializer="lecun_normal",
                                 activation="selu"))

model.add(keras.layers.AlphaDropout(rate=0.2))
model.add(keras.layers.Dense(10, activation="softmax"))

optimizer = keras.optimizers.SGD(learning_rate=0.000422)
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

In [73]:
batch_size = 128
n_epochs = 25
onecycle = OneCycleScheduler(math.ceil(len(X_train_scaled) / batch_size) * n_epochs, max_rate=0.05)
history = model.fit(X_train_scaled, y_train, epochs=n_epochs, batch_size=batch_size,
                    validation_data=(X_valid_scaled, y_valid),
                    callbacks=[onecycle])

Epoch 1/25
98/98 [==============================] - 5s 48ms/step - loss: 0.4806 - accuracy: 0.8316 - val_loss: 3.0436 - val_accuracy: 0.4378
Epoch 2/25
98/98 [==============================] - 3s 26ms/step - loss: 0.5122 - accuracy: 0.8183 - val_loss: 3.1088 - val_accuracy: 0.4355
Epoch 3/25
98/98 [==============================] - 3s 27ms/step - loss: 0.5980 - accuracy: 0.7923 - val_loss: 2.9893 - val_accuracy: 0.4325
Epoch 4/25
98/98 [==============================] - 3s 26ms/step - loss: 0.6939 - accuracy: 0.7579 - val_loss: 2.8764 - val_accuracy: 0.4311
Epoch 5/25
98/98 [==============================] - 3s 26ms/step - loss: 0.7290 - accuracy: 0.7410 - val_loss: 2.8321 - val_accuracy: 0.4160
Epoch 6/25
98/98 [==============================] - 3s 27ms/step - loss: 0.8117 - accuracy: 0.7159 - val_loss: 2.7948 - val_accuracy: 0.4219
Epoch 7/25
98/98 [==============================] - 3s 30ms/step - loss: 0.8264 - accuracy: 0.7078 - val_loss: 2.5387 - val_accuracy: 0.4100
Epoch 8/25
98

In [74]:
model.evaluate(X_test_scaled, y_test)

313/313 [==============================] - 1s 4ms/step - loss: 4.6512 - accuracy: 0.4491


[4.65122127532959, 0.44909998774528503]

* The training epoch is much faster, each taking only around 4 seconds. The test accuracy is around 45%, which is slightly slower than Batch Normalization.